source/common/init/target_impl.cc

```diff
@@ -50,5 +52,37 @@ bool TargetImpl::ready() {
   return false;
 }
 
+SharedTargetImpl::SharedTargetImpl(absl::string_view name, InitializeFn fn)
+    : name_(fmt::format("shared target {}", name)),
+      fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
+        if (initialized_) {
+          watcher_handle->ready();
+        } else {
+          watcher_handles_.push_back(std::move(watcher_handle));
+          std::call_once(once_flag_, fn);
+        }
+      })) {}
+
+SharedTargetImpl::~SharedTargetImpl() { ENVOY_LOG(debug, "{} destroyed", name_); }
+
+absl::string_view SharedTargetImpl::name() const { return name_; }
+
+TargetHandlePtr SharedTargetImpl::createHandle(absl::string_view handle_name) const {
+  // Note: can't use std::make_unique here because TargetHandleImpl ctor is private.
+  return absl::WrapUnique(
+      new TargetHandleImpl(handle_name, name_, std::weak_ptr<InternalInitalizeFn>(fn_)));
+}
+
+bool SharedTargetImpl::ready() {
+  initialized_ = true;
+  bool all_notified = true;
+  for (auto& watcher_handle : watcher_handles_) {
+    all_notified = watcher_handle->ready() && all_notified;
```

Comment:
Nit: `&=`

Reply:
Maybe not...
`watcher_handle->ready()` expression must be evaluated no matter all_notified is true or false.

Reply:
https://stackoverflow.com/questions/23107162/do-the-and-operators-for-bool-short-circuit